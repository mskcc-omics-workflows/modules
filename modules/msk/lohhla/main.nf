process LOHHLA {
    tag "$metaT.id"
    label 'process_high'
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'docker.io/orgeraj/lohhla:1.1.8':
        'docker.io/orgeraj/lohhla:1.1.8' }"

    input:
    tuple val(metaT), path(bamTumor)
    tuple val(metaN), path(bamNormal) 
    tuple val(metahlaN),path(winnersHla) //HLA output from polysolver or other hla caller. 
    tuple val(metaFacets) , path(purityOut)  // purityOut is from FACETS run
    
        

    output:
    tuple val(metaT), val(metaN), path("*.DNA.HLAlossPrediction_CI.txt"), path("*DNA.IntegerCPN_CI.txt"), path("*.pdf"), path("*.RData"), emit: lohhlaOutput
    path  "versions.yml"              , emit: versions


    when:
    task.ext.when == null || task.ext.when

    script:
    def args              = task.ext.args ?: ''
    def prefixTumor       = task.ext.prefix ?: "${metaT.id}"
    def prefixNormal      = task.ext.prefix ?: "${metaN.id}"
    def minCoverageFilter = 10
    def outputPrefix      = task.ext.prefix ?: "${prefixTumor}_${prefixNormal}"
    def hlaFasta          =  "/lohhla/data/abc_complete.fasta"
    def hlaDat            = "/lohhla/data/hla.dat"

    """

    samtools index -b ${bamTumor}
    samtools index -b ${bamNormal}

    cat ${winnersHla} | tr "\t" "\n" | grep -v "HLA" > massaged.winners.hla.txt

    PURITY=\$(grep Purity *_purity.out | grep -oP "[0-9\\.]+|NA+")
    [[ \$PURITY == "NA" ]] && PURITY=0.5
    PLOIDY=\$(grep Ploidy *_purity.out | grep -oP "[0-9\\.]+|NA+")
    cat <(echo -e "tumorPurity\ttumorPloidy") <(echo -e "${prefixTumor}\t\$PURITY\t\$PLOIDY") > tumor_purity_ploidy.txt

    Rscript --no-init-file /lohhla/LOHHLAscript.R \\
        --patientId ${outputPrefix} \\
        --normalBAMfile ${bamNormal} \\
        --tumorBAMfile ${bamTumor} \\
        --HLAfastaLoc ${hlaFasta} \\
        --HLAexonLoc ${hlaDat} \\
        --CopyNumLoc tumor_purity_ploidy.txt \\
        --minCoverageFilter ${minCoverageFilter} \\
        --hlaPath massaged.winners.hla.txt \\
        --gatkDir /picard-tools \\
        --novoDir /opt/conda/bin \\
        $args 

    if [[ -f ${outputPrefix}.${minCoverageFilter}.DNA.HLAlossPrediction_CI.txt ]]
    then
        sed -i "s/^${prefixTumor}/${outputPrefix}/g" ${outputPrefix}.${minCoverageFilter}.DNA.HLAlossPrediction_CI.txt
    else
        rm -rf *.DNA.HLAlossPrediction_CI.txt
    fi

    if [[ -f ${outputPrefix}.${minCoverageFilter}.DNA.IntegerCPN_CI.txt ]]
    then
        sed -i "s/^/${outputPrefix}\t/g" ${outputPrefix}.${minCoverageFilter}.DNA.IntegerCPN_CI.txt
        sed -i "0,/^${outputPrefix}\t/s//sample\t/" ${outputPrefix}.${minCoverageFilter}.DNA.IntegerCPN_CI.txt
    else
        rm -rf *.DNA.IntegerCPN_CI.txt
    fi

    touch ${outputPrefix}.${minCoverageFilter}.DNA.HLAlossPrediction_CI.txt
    touch ${outputPrefix}.${minCoverageFilter}.DNA.IntegerCPN_CI.txt

    mv ${outputPrefix}.${minCoverageFilter}.DNA.HLAlossPrediction_CI.txt ${outputPrefix}.DNA.HLAlossPrediction_CI.txt
    mv ${outputPrefix}.${minCoverageFilter}.DNA.IntegerCPN_CI.txt ${outputPrefix}.DNA.IntegerCPN_CI.txt

    if find Figures -mindepth 1 | read
    then
        mv Figures/* .
        mv ${prefixTumor}.minCoverage_${minCoverageFilter}.HLA.pdf ${outputPrefix}.HLA.pdf
    fi

    

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        lohhla: v1.17
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefixTumor       = task.ext.prefix ?: "${metaT.id}"
    def prefixNormal      = task.ext.prefix ?: "${metaN.id}"
    def minCoverageFilter = 10
    def outputPrefix      = task.ext.prefix ?: "${prefixTumor}_${prefixNormal}"
    """
    


    touch ${outputPrefix}.DNA.HLAlossPrediction_CI.txt
    touch ${outputPrefix}.DNA.IntegerCPN_CI.txt
    touch ${outputPrefix}.HLA.pdf

    my_array=("a" "b" "c")

    for hlagene in "\${my_array[@]}"
    do
        touch ${outputPrefix}.hla_\${hlagene}.tmp.data.plots.RData
    done

    touch ${outputPrefix}.hla_a.tmp.data.plots.RData

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        lohhla: v1.17
    END_VERSIONS
    """
}
