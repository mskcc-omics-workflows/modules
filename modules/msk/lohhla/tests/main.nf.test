nextflow_process {

    name "Test Process LOHHLA"
    script "../main.nf"
    process "LOHHLA"

    tag "modules"
    tag "modules_msk"
    tag "lohhla"
    tag "polysolver"
    tag "facets"
    tag "snppileup"

    // TODO nf-core: Change the test name preferably indicating the test-data and file-format used

    // TODO nf-core: If you are created a test for a chained module
    // (the module requires running more than one process to generate the required output)
    // add the 'setup' method here.
    // You can find more information about how to use a 'setup' method in the docs (https://nf-co.re/docs/contributing/modules#steps-for-creating-nf-test-for-chained-modules).

    test("homo_sapiens - bam") {
        setup {
            run("SNPPILEUP") {
                script "../../snppileup/main.nf"
                config "../../snppileup/tests/nextflow.config"
                process {
                    """
                        meta = [ id:'test', single_end:false ] // meta map
                        input[0] = [ meta,
                                    //file(params.test_data['homo_sapiens']['illumina']['test_paired_end_hla_sorted_bam'], checkIfExists: true),
                                    //file(params.test_data['homo_sapiens']['illumina']['test_paired_end_hla_sorted_bam_bai'], checkIfExists: true),
                                    //file(params.test_data['homo_sapiens']['illumina']['test_rna_paired_end_sorted_chr6_bam'], checkIfExists: true),
                                    //file(params.test_data['homo_sapiens']['illumina']['test_rna_paired_end_sorted_chr6_bam_bai'], checkIfExists: true),
                                    file(params.test_data_mskcc['hla_bams']['test_hlaregion_bam'], checkIfExists: true),
                                    file(params.test_data_mskcc['hla_bams']['test_hlaregion_bam_bai'], checkIfExists: true),
                                    file(params.test_data_mskcc['hla_bams']['test_tumor_hlaregion_bam'], checkIfExists: true),
                                    file(params.test_data_mskcc['hla_bams']['test_tumor_hlaregion_bam_bai'], checkIfExists: true),

                                    [],
                                    []
                                ]
                        input[1] = [
                                        meta,
                                        file(params.test_data_mskcc['hla_bams']['dbsnp_hla_chr_vcf_gz'], checkIfExists: true),
                                        file(params.test_data_mskcc['hla_bams']['dbsnp_hla_chr_vcf_gz_tbi'], checkIfExists: true)
                                ]
                    """
                }
            }

            run("FACETS") {
                script "../../facets/main.nf"
                config "../../facets/tests/nextflow.config"
                process {
                    """
                    meta = [ id:'test_hla', single_end:false ] // meta map
                    input[0] = SNPPILEUP.out.pileup.collect{ meta, pileup -> pileup }.map{ pileup -> [meta, pileup, true]}
                    """
                }
            }

            run("POLYSOLVER") {
                script "../../polysolver/main.nf"
                process {
                    """
                    input[0] = [
                        [ id:'test_hla', single_end:false ],
                        file(params.test_data_mskcc['hla_bams']['test_hlaregion_bam'], checkIfExists: true)
                    ]
                    input[1] = []
                    input[2] = []
                    input[3] = []
                    input[4] = []

                    """
                }
            }
        }
        when {
            process {
                """
                meta = [ id:'test_hla', single_end:false ]
                input[0] = Channel.of([
                    meta,
                    file(params.test_data_mskcc['hla_bams']['test_tumor_hlaregion_bam'], checkIfExists: true),
                    file(params.test_data_mskcc['hla_bams']['test_hlaregion_bam'], checkIfExists: true),
                ]).join(POLYSOLVER.out.hla, by:0)
                .join(FACETS.out.purity_out, by:0)
                """


            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
                //TODO nf-core: Add all required assertions to verify the test output.
            )
        }

    }
    test("stub") {
        options "-stub-run"
        when {
            process {
                """
                input[0] = [[ id:'test_hla'],[],[],[],[]]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

}
