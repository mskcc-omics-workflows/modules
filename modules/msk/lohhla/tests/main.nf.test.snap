{
    "stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_hla"
                        },
                        "test_hla__test_hla.DNA.HLAlossPrediction_CI.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_hla"
                        },
                        "test_hla__test_hla.DNA.IntegerCPN_CI.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test_hla"
                        },
                        "test_hla__test_hla.HLA.pdf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test_hla"
                        },
                        [
                            "test_hla__test_hla.hla_a.tmp.data.plots.RData:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_hla__test_hla.hla_b.tmp.data.plots.RData:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_hla__test_hla.hla_c.tmp.data.plots.RData:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "4": [
                    "versions.yml:md5,f3caa23f6a246ce5fdc3021c31ccc5ae"
                ],
                "integercpn": [
                    [
                        {
                            "id": "test_hla"
                        },
                        "test_hla__test_hla.DNA.IntegerCPN_CI.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "pdf": [
                    [
                        {
                            "id": "test_hla"
                        },
                        "test_hla__test_hla.HLA.pdf:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "prediction": [
                    [
                        {
                            "id": "test_hla"
                        },
                        "test_hla__test_hla.DNA.HLAlossPrediction_CI.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "rdata": [
                    [
                        {
                            "id": "test_hla"
                        },
                        [
                            "test_hla__test_hla.hla_a.tmp.data.plots.RData:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_hla__test_hla.hla_b.tmp.data.plots.RData:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_hla__test_hla.hla_c.tmp.data.plots.RData:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f3caa23f6a246ce5fdc3021c31ccc5ae"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-12T18:21:59.059169"
    },
    "homo_sapiens - bam": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_hla",
                            "single_end": false
                        },
                        "test_hla__test_hla.DNA.HLAlossPrediction_CI.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test_hla",
                            "single_end": false
                        },
                        "test_hla__test_hla.DNA.IntegerCPN_CI.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,aaaf018a5ada92678b89b6497c013ab6"
                ],
                "integercpn": [
                    [
                        {
                            "id": "test_hla",
                            "single_end": false
                        },
                        "test_hla__test_hla.DNA.IntegerCPN_CI.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "pdf": [
                    
                ],
                "prediction": [
                    [
                        {
                            "id": "test_hla",
                            "single_end": false
                        },
                        "test_hla__test_hla.DNA.HLAlossPrediction_CI.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "rdata": [
                    
                ],
                "versions": [
                    "versions.yml:md5,aaaf018a5ada92678b89b6497c013ab6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-12T18:21:51.776514"
    }
}